lib/runner.js

```diff
@@ -0,0 +1,175 @@
+'use babel';
+
+import { Emitter, BufferedProcess } from 'atom';
+import fs from 'fs';
+import path from 'path';
+
+export default class Runner {
+
+  // Public: Creates a Runner instance
+  //
+  // * `scriptOptions` a {ScriptOptions} object instance
+  // * `emitter` Atom's {Emitter} instance. You probably don't need to overwrite it
+  constructor(scriptOptions) {
+    this.bufferedProcess = null;
+    this.stdoutFunc = this.stdoutFunc.bind(this);
+    this.stderrFunc = this.stderrFunc.bind(this);
+    this.onExit = this.onExit.bind(this);
+    this.createOnErrorFunc = this.createOnErrorFunc.bind(this);
+    this.scriptOptions = scriptOptions;
+    this.emitter = new Emitter();
+  }
+
+  run(command, extraArgs, codeContext, inputString = null) {
+    this.startTime = new Date();
+
+    const args = this.args(codeContext, extraArgs);
+    const options = this.options();
+    const stdout = this.stdoutFunc;
+    const stderr = this.stderrFunc;
+    const exit = this.onExit;
+
+    this.bufferedProcess = new BufferedProcess({
+      command, args, options, stdout, stderr, exit,
+    });
+
+    if (inputString) {
+      this.bufferedProcess.process.stdin.write(inputString);
+      this.bufferedProcess.process.stdin.end();
+    }
+
+    this.bufferedProcess.onWillThrowError(this.createOnErrorFunc(command));
+  }
+
+  stdoutFunc(output) {
+    this.emitter.emit('did-write-to-stdout', { message: output });
+  }
+
+  onDidWriteToStdout(callback) {
+    return this.emitter.on('did-write-to-stdout', callback);
+  }
+
+  stderrFunc(output) {
+    this.emitter.emit('did-write-to-stderr', { message: output });
+  }
+
+  onDidWriteToStderr(callback) {
+    return this.emitter.on('did-write-to-stderr', callback);
+  }
+
+  destroy() {
+    this.emitter.dispose();
+  }
+
+  getCwd() {
+    let cwd = this.scriptOptions.workingDirectory;
+
+    if (cwd) {
+      switch (atom.config.get('script.cwdBehavior')) {
+        case 'First project directory': {
+          const paths = atom.project.getPaths();
+          if (paths && paths.length > 0) {
+            try {
+              cwd = fs.statSync(paths[0]).isDirectory() ? paths[0] : path.join(paths[0], '..');
+            } catch (error) { /* Don't throw */ }
+          }
+          break;
+        }
+        case 'Project directory of the script': {
+          cwd = this.getProjectPath();
+          break;
+        }
+        case 'Directory of the script': {
+          const pane = atom.workspace.getActivePaneItem();
+          cwd = (pane && pane.buffer && pane.buffer.file && pane.buffer.file.getParent &&
+                 pane.buffer.file.getParent() && pane.buffer.file.getParent().getPath &&
+                 pane.buffer.file.getParent().getPath()) || '';
```

Comment:
Can really everything fail?

Before it was:
```
cwd = atom.workspace.getActivePaneItem()?.buffer?.file?.getParent?().getPath?() or ''
```
That's the one thing ES6 is missing 😄